### ISSUE TITLE
Stuck on waiting for video connection in test_camera.py

### ISSUE THREAD

**Issue by jashshah999** [kripper/SmartLanding#2]
I followed all the steps and rebuilt the rosetta app using instructions from hacking on rosetta and installed it on my phone and then changed the IP along with the port in rosetta_camera.py. I then ran python3 test_camera.py and it is stuck on waiting for connection.. 

Am I missing something? Also there is a step to change something in vision.conf but I cannot find this file anywhere. 

I am trying to get a video stream from my DJI Mini SE to a dronekit script (maybe even convert it to a ros topic). But I am not able to do this since a few days. Please help me out. Thanks.

**Comment by jashshah999** [comment]
I did not set the IP correctly in the app. Because I am using a local network, IP keeps changing so I need to switch it in the app again and again. Anyhow, I got that working but there is a weird video stream being published 

![image](https://github.com/RosettaDrone/SmartLanding/assets/49280550/1d6c9d18-8961-4fa1-8f6f-7e8681624f4c)

**Comment by kripper** [comment]
This is the test mode. In real mode, you should be able to see the camera stream. Or you could also change the variable in the code.

**Comment by jashshah999** [comment]
I did it in both test and real mode, I am getting the same output. Which variable are you exactly talking about? What do I change? 

Until now, I changed the Rosetta code as you had instructed on "hacking on Rosetta". I didn't find the vision.conf file  in smart landing so didn't make any tcp change there. 

I then rebuilt the apk. Now my video feed is gone from Rosetta app as expected. But I'm not able to see anything on this window. Am I missing something?

**Comment by kripper** [comment]
https://github.com/RosettaDrone/rosettadrone/blob/86fa0e566a7c24abca19db9fa0099ef75341e1db/app/src/main/java/sq/rogue/rosettadrone/plugins/RawVideoStreamer.java#L29

**Comment by kripper** [comment]
BTW, I'm assuming you are using the plug-in to do Vision Landing.

**Comment by kripper** [comment]
BTW, I'm assuming you are using the plug-in to do Vision Landing.

**Comment by jashshah999** [comment]
I'm getting this error after changing the flag in the line you mentioned. I
think there's a shape issue.

yuv_mat = yuv_mat.reshape((h + h // 2, w))
ValueError: cannot reshape array of size 1384448 into shape (1080,1280)

Regards,
Jash Shah

**Comment by kripper** [comment]
Same as https://github.com/RosettaDrone/rosettadrone/issues/151

**Comment by jashshah999** [comment]
So shall I directly use the other repo which is linked in the end of that issue? Would that solve my problem?

**Comment by kripper** [comment]
Please check related issue:

`To fix, you should check the source code that grabs and sends the video frame and find out what format it is using, in order to decode it correctly on the receiving end (vision landing).
`

**Comment by jashshah999** [comment]
But for my mini se, should I use the repo linked or should I use the same
Rosetta drone repo and change the python code decoding? My question is
should I use the same Rosetta and change the python decoding or keep the
python decoding the same and use the other repo? I mean should I also make those changes which are on hacking on Rosetta or no?

**Comment by kripper** [comment]
Always use the latest GitHub version and follow the hacking guide.

**Comment by jashshah999** [comment]
Correct. I did that but faced the shape issue. Do you recommend I use this repo instead (https://github.com/m4xw/rosettadrone_mini2) or is the main branch on rosetta drone good enough with the changes made using hacking guide?

**Comment by kripper** [comment]
I don't think so. It's an old unsupported bunch of hacks to support mini 2 before it was officially supported by DJI.

**Comment by jashshah999** [comment]
Okay but then I am confused as to what I should do to get the feed because I clone the latest rosetta_drone and did the steps as required. I have DJI Mini SE and I am igonring the last 2048 bytes coming in like this in the decoding script: 
`yuv_mat = np.frombuffer(buffer, dtype=np.uint8)[:-2048]`

Now, I finally got a video feed but its extremely laggy. What can I do to fix this, is there an issue already for laggy stream from DJI mini se?

**Comment by jashshah999** [comment]
The lag is a few seconds. Is there anything I can do to fix it?

**Comment by jashshah999** [comment]
[Screencast from 08-13-2023 06:49:37 PM.webm](https://github.com/RosettaDrone/SmartLanding/assets/49280550/dfca0cb6-02d5-431b-a87a-8ba6100d4ed2)

Current state:

I am using rosetta_drone's most recent version with all the changes mentioned in hacking with rosetta except the vision.conf file which I could not find. I also made a change in the decoding by using everything exceept the last 2048 bytes by doing the below line. 

`yuv_mat = np.frombuffer(buffer, dtype=np.uint8)[:-2048]`

Am I missing something here? Or doing something wrong?

**Comment by jashshah999** [comment]
Really need help with this. Please let me know what needs to be done. Thanks

**Comment by kripper** [comment]
Check if the sent frame size is exactly the same as the expected size of the receiver.

Also check if your hardware/network is fast enough to process the frames.

You could also reduce the frames per second.

The code is quite simple.

Your images are distorted. Maybe you are not using a fan to cool the drone.

**Comment by jashshah999** [comment]
Hey Christopher,

I tried all of the above and I am still facing issues.

1. The frame size sent and received is exactly the same (1280*720)
2. I changed the network to a faster WiFi. I did see some improvement but
marginally. I am attaching a video for the same.
3. I reduced frames per second, from 30 to 15 to even 1 but did not see
significant change.
4. I also used a fan to cool the drone but it did not help at all.

My main concern is that when I am testing outdoors, I will not have a fan
to cool the drone or even maybe such high speed wifi, I might have to use a
hotspot when I am testing outdoors. Is there a workaround for this? I am
stuck and not able to proceed with my project because of this. It is worth
noting that when the data was coming previously via UDP, QGC was perfectly
able to decode it. Yes, there was some latency but it was negligible
compared to the latency I am seeing now. Is there a way I can maybe revert
to the UDP version and use gstreamer to view my data and convert it to a
ROS topic which can be used to perform other aspects of my project. Any
help is appreciated. Thank you,
Regards,
Jash Shah

**Comment by kripper** [comment]
> My main concern is that when I am testing outdoors, I will not have a fan
to cool the drone

A fan is not necessary during flight, as the airflow naturally cools down the drone.

> or even maybe such high speed wifi, I might have to use a hotspot when I am testing outdoors

Does the Wi-Fi connect your smartphone to your PC? What are you running on the PC? I run everything directly on the smartphone. However, I believe you might have another issue that is not related to network speed.

>  I reduced frames per second, from 30 to 15 to even 1 but did not see significant change.

Keep it on 1 FPS until you have found and fixed the problem.

> It is worth noting that when the data was coming previously via UDP, QGC was perfectly able to decode it.

Interesting. Are you sure you have a "DJI Mini SE"?
The issue you mentioned before (https://github.com/RosettaDrone/rosettadrone/issues/151) doesn't occur on a "DJI Mini SE".
Regarding QGC, please see and comment here: https://github.com/RosettaDrone/rosettadrone/issues/174

> Is there a way I can maybe revert to the UDP version and use gstreamer to view my data and convert it to a ROS topic

Yes, but we must first get the "DJI Mini SE" RTP Video Streaming working fine with GStreamer.

Please note that QGC and Gstreamer both use RTP video streaming, while the RawVideoPlugin (`test_camera.py`) sends raw image frames.

**Comment by jashshah999** [comment]
Hey, yes that is my issue itself. I raised the QGC issue there and the python script issue here. We can continue the discussion there to debug gstreamer first if required.
